include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,374 @@
+ *       derived from this software without specific prior written permission.
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+ * ARE DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef _ONEDPL_SCAN_BY_SEGMENT_IMPL_H
+#define _ONEDPL_SCAN_BY_SEGMENT_IMPL_H
+
+#if _ONEDPL_BACKEND_SYCL
+
+#include <oneapi/dpl/pstl/algorithm_fwd.h>
+#include <oneapi/dpl/pstl/parallel_backend.h>
+#include <oneapi/dpl/pstl/hetero/utils_hetero.h>
+
+#include <oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h>
+#include <oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h>
+#include <oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h>
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace internal
+{
+
+// This function is responsible for performing an exclusive segmented scan between work items in shared local memory.
+// Each work item passes __delta_local_id which is the distance to the closest lower indexed work item that
+// processes a segment end. A SLM accessor that is twice the work-group size must be provided.
+// Example across eight work items with addition as the binary operation :
+// -----------------------------
+// Local ID    : 0 1 2 3 4 5 6 7
+// Accumulator : 2 3 1 3 0 5 2 1
+// Delta       : 0 1 2 3 0 1 0 0
+// -----------------------------
+// Return Vals : 0 2 5 6 9 0 5 2
+// -----------------------------
+template <typename _NdItem, typename _LocalAcc, typename _IdxType, typename _ValueType, typename _BinaryOp>
+inline _ValueType
+wg_segmented_scan(_NdItem __item, _LocalAcc __local_acc, _IdxType __local_id, _IdxType __delta_local_id,
+                  _ValueType __accumulator, _ValueType __identity, _BinaryOp __binary_op, ::std::size_t __wgroup_size)
+{
+    _IdxType __first = 0;
+    __local_acc[__local_id] = __accumulator;
+
+    __dpl_sycl::__group_barrier(__item);
+
+    for (::std::size_t __i = 1; __i < __wgroup_size; __i *= 2)
+    {
+        if (__delta_local_id >= __i)
+            __accumulator = __binary_op(__local_acc[__first + __local_id - __i], __accumulator);
+
+        __first = __wgroup_size - __first;
+        __local_acc[__first + __local_id] = __accumulator;
+        __dpl_sycl::__group_barrier(__item);
+    }
+
+    return (__local_id ? __local_acc[__first + __local_id - 1] : __identity);
+}
+
+enum class scan_type
```

Comment:
I am wondering why it's done via `enum` while further it's used as `bool` (if I am not missing something)

Reply:
I just did this for stylistic reasons so its defined as either `sycl_scan_by_segment_impl<scan_type::inclusive> __scan` or `sycl_scan_by_segment_impl<scan_type::exclusive> __scan`.

I think changing it to a bool is more consistent with how we write  `sycl_scan_by_segment_impl</*is_inclusive=*/true> __scan` code in other areas, so I can change it to this.

Reply:
I have made this change.